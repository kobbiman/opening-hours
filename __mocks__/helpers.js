/* eslint-disable react/prop-types */
import React from 'react'
import { ThemeProvider } from 'styled-components'
import { theme } from '@lib/theming'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { render } from '@testing-library/react'

export const testRenderer = el => renderer.create(el).toJSON()

export const ThemedComponent = ({ children }) => <ThemeProvider theme={theme}>{children}</ThemeProvider>

export const renderWithTheme = component => <ThemedComponent>{component}</ThemedComponent>

export const toHaveStyleRule = (component, property, value, options) => {
  expect(testRenderer(component)).toHaveStyleRule(property, value, options)
  expect(mount(component)).toHaveStyleRule(property, value, options)
  expect(render(component).container.firstChild).toHaveStyleRule(property, value, options)
}

export const shouldRenderProp = (component, as = 'div') => {
  const Component = component
  const id = 'testId'
  const { container } = render(<Component id={id} as={as} />)
  expect(container.querySelector(as).id).toEqual(id)
}

export const shouldRenderSomeText = (component, as = 'div') => {
  const Component = component
  const children = 'Text'
  const { container, queryByText } = render(<Component as={as}>{children}</Component>)
  const { childNodes } = container.querySelector(as)
  expect(childNodes).toHaveLength(1)
  expect(queryByText(children)).not.toBeNull()
}

export const mockRawOpeningHours = {
  monday: [],
  tuesday: [],
  wednesday: [],
  thursday: [],
  friday: [
    {
      type: 'open',
      value: 64800,
    },
  ],
  saturday: [
    {
      type: 'close',
      value: 3600,
    },
    {
      type: 'open',
      value: 32400,
    },
    {
      type: 'close',
      value: 39600,
    },
    {
      type: 'open',
      value: 57600,
    },
    {
      type: 'close',
      value: 86399,
    },
  ],
}

export const mockFormattedOpeningHours = [
  { key: 'monday', value: [] },
  { key: 'tuesday', value: [] },
  { key: 'wednesday', value: [] },
  { key: 'thursday', value: [] },
  { key: 'friday', value: [['6 PM', '1 AM']] },
  {
    key: 'saturday',
    value: [
      ['9 AM', '11 AM'],
      ['4 PM', '11:59 PM'],
    ],
  },
]
