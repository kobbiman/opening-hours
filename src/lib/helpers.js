import PropTypes from 'prop-types'

const SECONDS_IN_HOUR = 3600

export const formatTime = value => {
  const hours = Math.floor(value / SECONDS_IN_HOUR)
  const minutes = Math.floor((value % SECONDS_IN_HOUR) / 60)

  const time = [hours % 12 || 12]

  if (minutes > 0) {
    time.push(`0${minutes}`.slice(-2))
  }

  const amOrPm = hours < 12 ? 'AM' : 'PM'
  return [time.join(':'), amOrPm].join(' ')
}

/**
 *
 * @param {Object} dayValues
 * @param {Object} nextDayValues
 */

const parseDay = (dayValues, nextDayValues) => {
  if (dayValues.length === 0) return dayValues

  const entries = []

  for (let index = 0; index < dayValues.length; index++) {
    const { value, type } = dayValues[index]
    const entryTime = formatTime(value)

    if (type === 'open') {
      // Add a new sub array if it's not the last entry
      if (index < dayValues.length - 1) {
        entries.push([entryTime])
      } else {
        // If it's 'open' and the last entry , find first 'close' in the next day
        entries.push([entryTime, formatTime(nextDayValues[0].value)])
      }
    } else {
      // If the first value of the array is 'close', do nothing as its irrelevant for the day context, otherwise
      // add entry value to last sub array in array
      if (index !== 0) {
        entries[entries.length - 1].push(entryTime)
      }
    }
  }

  return entries
}

export const formatOpeningHours = data => {
  if (!data) return []
  const keys = Object.keys(data)
  const formattedDays = keys.map((key, i) => {
    const dayData = data[key]

    // Get next day value, if index out of bounds return first item
    const nextDay = data[keys[i + 1]] || data[keys[0]]
    return { key, value: parseDay(dayData, nextDay) }
  })

  return formattedDays
}

export const weekdayName = new Date().toLocaleDateString('en', { weekday: 'long' }).toLowerCase()

export const TuppleChecker = (...validators) =>
  PropTypes.arrayOf((_, index) => {
    const currentValidators = validators.filter((v, i) => i === index)
    if (currentValidators.length <= 0) return true
    const [currentValidator] = currentValidators

    return currentValidator
  })
