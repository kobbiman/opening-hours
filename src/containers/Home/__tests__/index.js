import React from 'react'
import { render } from '@testing-library/react'
import { ThemedComponent, mockRawOpeningHours, mockFormattedOpeningHours } from '__mocks__/helpers'
import { formatOpeningHours } from '@lib/helpers'

import HomePage from '../index'

export const mockInitialProps = overrides => {
  const parsedOpeningHours = formatOpeningHours(mockRawOpeningHours)
  return {
    openingHours: parsedOpeningHours,
    ...overrides,
  }
}

const mockGetInitialProps = jest.fn(mockInitialProps)

describe('<HomePage />', () => {
  it('should render as expected with passed component and props', async () => {
    HomePage.getInitialProps = mockGetInitialProps
    const props = await HomePage.getInitialProps()

    const {
      container: { firstChild },
    } = render(
      <ThemedComponent>
        <HomePage {...props} />
      </ThemedComponent>,
    )

    expect(firstChild).toMatchSnapshot()
    expect(props).toEqual({
      openingHours: mockFormattedOpeningHours,
    })
    expect(HomePage.getInitialProps).toHaveBeenCalledTimes(1)
  })
})
