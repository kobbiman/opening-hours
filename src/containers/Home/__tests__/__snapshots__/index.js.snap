// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render as expected with passed component and props 1`] = `
.c5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-weight: 700;
  font-size: 24px;
  line-height: 30px;
  margin-left: 12px;
}

.c8 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
  font-weight: 500;
  line-height: 22px;
}

.c11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: #A1A2A4;
}

.c12 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-weight: 700;
  font-size: 12px;
  text-transform: uppercase;
  line-height: 16px;
  color: #5BCB02;
  margin-left: 8px;
}

.c13 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.c7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c10 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  background-color: #ffffff;
  box-shadow: 1px 1px 6px 1px rgba(0,0,0,0.1);
  border-radius: 12px;
  padding: 25px 25px 30px;
  width: 100%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-bottom: 2px solid #696a6c;
  padding-bottom: 12px;
  margin-bottom: 2px;
}

.c4 {
  width: 20px;
  height: 20px;
  color: #A1A2A4;
}

.c6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-bottom: 1px solid #EEEEEE;
  padding: 8px 0;
}

.c9 {
  text-transform: capitalize;
}

.c0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: auto;
  margin-right: auto;
  padding-left: 12px;
  padding-right: 12px;
  width: 100%;
}

.c1 {
  height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width:768px) {
  .c2 {
    width: 345px;
  }
}

@media (min-width:576px) {
  .c0 {
    max-width: 540px;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 720px;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: 960px;
  }
}

@media (min-width:1280px) {
  .c0 {
    max-width: 1264px;
  }
}

<div
  class="c0 c1"
  width="100%"
>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <svg
        class="c4"
        height="1em"
        viewBox="15 20 70 60"
        width="1em"
      >
        <path
          d="M50 17c-18.202 0-33 14.798-33 33 0 18.202 14.798 33 33 33 18.202 0 33-14.798 33-33 0-18.202-14.798-33-33-33zm0 4c16.04 0 29 12.96 29 29S66.04 79 50 79 21 66.04 21 50s12.96-29 29-29zm0 8a2 2 0 0 0-2 2v19c0 .552.232 1.044.594 1.406l12 12.032a2.018 2.018 0 0 0 2.843 0 2.019 2.019 0 0 0 0-2.844L52 49.188V31a2 2 0 0 0-2-2z"
          fill="currentColor"
        />
      </svg>
      <h1
        class="c5"
      >
        Opening hours
      </h1>
    </div>
    <div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8 c9"
          >
            monday
          </div>
        </div>
        <div
          class="c10"
        >
          <div
            class="c11"
            color="grey3"
          >
            Closed
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8 c9"
          >
            tuesday
          </div>
          <h5
            class="c12"
            color="green"
          >
            today
          </h5>
        </div>
        <div
          class="c10"
        >
          <div
            class="c11"
            color="grey3"
          >
            Closed
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8 c9"
          >
            wednesday
          </div>
        </div>
        <div
          class="c10"
        >
          <div
            class="c11"
            color="grey3"
          >
            Closed
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8 c9"
          >
            thursday
          </div>
        </div>
        <div
          class="c10"
        >
          <div
            class="c11"
            color="grey3"
          >
            Closed
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8 c9"
          >
            friday
          </div>
        </div>
        <div
          class="c10"
        >
          <div
            class="c13"
          >
            <span>
              <span>
                6 PM - 1 AM
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8 c9"
          >
            saturday
          </div>
        </div>
        <div
          class="c10"
        >
          <div
            class="c13"
          >
            <span>
              <span>
                9 AM - 11 AM
              </span>
              <span>
                , 
              </span>
            </span>
            <span>
              <span>
                4 PM - 11:59 PM
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
